scanner/image/scanner/download-mappings.sh

```diff
@@ -20,11 +25,25 @@ urls=(
 for url in "${urls[@]}"; do
     filename=$(basename "$url")
     echo "Downloading ${url} > ${output_dir}/$filename"
-    curl --retry 3 -sSL --fail -o "${output_dir}/$filename" "$url"
+    curl --location --silent --fail --show-error --retry 3 --max-time 60 \
+        --output "${output_dir}/$filename" "$url"
     if [[ ! (-s "${output_dir}/$filename") ]]; then
```

Comment:
In general, these validations work as a stop-gap but what I [meant](https://gitlab.cee.redhat.com/stackrox/rhacs-midstream/-/merge_requests/3640#note_13500184) is deeper validation. Here we check that the files aren't empty and are pareseable JSONs but we don't check JSON structure and data is what we expect and so aren't protected against unexpected updates that break our parsers or missing data that breaks indexing or CVE matching.

It would not be feasible to add these deeper checks in bash or in Python. That is application logic and it should remain in the language of the application, Go.

The suggestion is: after placing these files in the Scanner V4 container, run our own scanner binary to load and check these files - in the same `Dockerfile`. This may require adding a new subcommand but shouldn't lead to writing duplicate code, only some refactorings. That way we can be more confident the files placed in the container are reasonably safe.

I'm not suggesting to do this now but I'd like you or Scanner team colleagues to confirm if it makes any sense and ideally create a backlog JIRA.

Reply:
I agree with your point regarding deeper validation. However, we should trust that the artifacts are correct. It's not much different than calling a library; I wouldn't write a test to ensure the library's correctness. Moreover, claircore doesn't perform any validation on this file and puts the contents in a Go map (AFAICT).

I suggest we ask ProdSec to provide a hash of these files and validate the files against those hashes.

In any case, it sounds like we want to solve this problem one way or another, so I'll create a ticket.

Reply:
> It's not much different than calling a library; I wouldn't write a test to ensure the library's correctness.

Some folks do exactly that and it gets them a long way in terms of resiliency to dependency updates and quality. </offtopic>

----

> However, we should trust that the artifacts are correct.

I think there is a thin line between these two cases:
* We downloaded the wrong thing because we downloaded the wrong way.
* We downloaded the wrong thing because it was indeed the wrong thing that we found and downloaded correctly.

Downloading and checking hashes protects us from partial downloads and the first case in general but it would not protect us from finding bogus data with a perfectly matching hash published by ProdSec by mistake or a failure. Everyone makes mistakes, that's in human nature.

Let's add a third case:
* ProdSec changes the format of these files in a way that breaks our logic.

We download correctly, hashes are valid, but the thing causes problems when used.
The similar in fact happened to NVD data (if I'm not mistaken) and caused outage.

Reply:
> > It's not much different than calling a library; I wouldn't write a test to ensure the library's correctness.
> Some folks do exactly that and it gets them a long way in terms of resiliency to dependency updates and quality.

No doubt that there are folks who do this and get safer updates. It's a balance between safe updates and velocity. For example, if we test every dependency in every component, we'll certainly have _no doubts_ that the dependencies work. But there's an opportunity cost to this. Instead of spending time on other work (e.g., performance, tech debt, features, etc), we have to spend time ensuring those tests are correct and maintained properly.

In this particular case, I'm not convinced it's worth it. If something is wrong with ProdSec's, we should solve the problem there rather than in our product (we're all Red Hat, after all 🙂).

That being said, I'm open-minded about this topic and could be convinced otherwise.

---

Commentary on the cases you outlined:

> We downloaded the wrong thing because we downloaded the wrong way.
> ...
> Downloading and checking hashes protects us from partial downloads and the first case in general...

Agreed.

> We downloaded the wrong thing because it was indeed the wrong thing that we found and downloaded correctly.
> ...
>  ... it would not protect us from finding bogus data with a perfectly matching hash published by ProdSec by mistake or a failure. Everyone makes mistakes, that's in human nature.

I agree with this sentiment. In this case, I believe it should be solved during runtime by claircore. In fact, I'd really like to solve this case in this ticket: [CLAIRDEV-99](https://issues.redhat.com/browse/CLAIRDEV-99).

> ProdSec changes the format of these files in a way that breaks our logic.

This is a _much_ larger problem to consider. This would break all scanners out in the wild and would require an update to all of them. Maybe I'm misunderstanding this case, but I'm not sure how exactly we should protect ourselves in this case.

Reply:
> This is a much larger problem to consider. This would break all scanners out in the wild

This is exactly what happened, and I found it. I hope the link works for you <https://redhat-internal.slack.com/archives/C0482EC01B2/p1666269636701089>. Further links are in ROX-13223 such as post-mortem docs, etc.

I hope by "larger" you did not mean "less probable".

Reply:
> > This is a much larger problem to consider. This would break all scanners out in the wild
>
> This is almost exactly what happened, and I found it. I hope the link works for you https://redhat-internal.slack.com/archives/C0482EC01B2/p1666269636701089. Further links are in [ROX-13223](https://issues.redhat.com/browse/ROX-13223) such as post-mortem docs, etc.

AFAICT, the only way for us to catch this issue during build time (and not runtime) is if we were to have a proper check for it, we build at an interval where central correctly has a version that works and doesn't yet have a broken version of the file, and we build during that interval.

> I hope by "larger" you did not mean "less probable".

Your hope is well-placed 🙂
I indeed did not mean "less probable"; I meant it would be much more catastrophic (and a problem I'm not convinced we could reasonably catch during build time)

Reply:
From what I can tell, we are all onboard with adding even more validation, but at a later time. I think the safeguards we have now are good for our immediate need. If a ticket doesn't already exist, let's make one (does one already exist? If not, I can make it)

Reply:
@RTann I haven't seen the ticket, so please create it.

@BradLugo I don't understand the suggestion.

> AFAICT, the only way for us to catch this issue during build time (and not runtime) is if we were to have a proper check for it, we build at an interval where central correctly has a version that works and doesn't yet have a broken version of the file, and we build during that interval.

Are you suggesting to have a running system (central?) which tells us if the files it recently downloaded are valid and we consult it to assess whether the builds produced during that time include valid files?

Reply:
> and perhaps not something we can solve in this PR.

This was my point from the very beginning: (a) let's be clear to ourselves that we aren't solving all problems that can occur around **this piece** of code, (b) let's acknowledge the problems aren't imaginary and backlog an item to deal with them.

Reply:
Ticket for follow-up work: https://issues.redhat.com/browse/ROX-26993